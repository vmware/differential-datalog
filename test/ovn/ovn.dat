start;

insert ovn_nb.NB_Global(0, 1, 2, 3, types.None);
insert ovn_nb.NB_Global(0, 1, 2, 3, types.Some{0x2cde1d4c89df430a9694714eb059969a});

commit;

dump;

exit;
