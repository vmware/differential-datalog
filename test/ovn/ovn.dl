extern type eth_addr

extern function eth_addr_zero(): eth_addr
extern function eth_addr2string(addr: eth_addr): string

function ip_fmt(ip: ovs_be32): string = {
    $"${ntohl(ip) >> 32'd24}.${(ntohl(ip) >> 32'd16) & 32'hff}.${(ntohl(ip) >> 32'd8) & 32'hff}.${ntohl(ip) & 32'hff}"
}

extern type in6_addr

typedef ovs_be32 = bit<32>

typedef ipv4_netaddr = IPV4NetAddr {
    addr: ovs_be32,            /* 192.168.10.123 */
    mask: ovs_be32,            /* 255.255.255.0 */
    network: ovs_be32,         /* 192.168.10.0 */
    plen: bit<32>,             /* CIDR Prefix: 24. */

    addr_s: string,            /* "192.168.10.123" */
    network_s: string,         /* "192.168.10.0" */
    bcast_s: string            /* "192.168.10.255" */
}

typedef ipv6_netaddr = IPV6NetAddr {
    addr: in6_addr,          /* fc00::1 */
    mask: in6_addr,          /* ffff:ffff:ffff:ffff:: */
    sn_addr: in6_addr,       /* ff02:1:ff00::1 */
    network: in6_addr,       /* fc00:: */
    plen: bit<32>,           /* CIDR Prefix: 64 */

    addr_s: string,          /* "fc00::1" */
    sn_addr_s: string,       /* "ff02:1:ff00::1" */
    network_s: string        /* "fc00::" */
}

typedef lport_addresses = LPortAddress {
    ea_s: string,
    ea: eth_addr,
    ipv4_addrs: Vec<ipv4_netaddr>,
    ipv6_addrs: Vec<ipv6_netaddr>
}

extern function warn(msg: string): ()

/*
 * C functions imported from OVN
 */
extern function is_dynamic_lsp_address(addr: string): bool
extern function extract_lsp_addresses(address: string): Option<lport_addresses>
extern function split_addresses(addr: string): (Set<string>, Set<string>)

/*
 * C functions imported from OVS
 */
extern function in6_generate_lla(ea: eth_addr): in6_addr

extern function ipv6_string_mapped(addr: in6_addr): string
extern function ipv6_parse_masked(s: string): Either<string/*err*/, (in6_addr/*ip*/, in6_addr/*mask*/)>
extern function ipv6_addr_bitxor(a: in6_addr, b: in6_addr): in6_addr
extern function ipv6_addr_bitand(a: in6_addr, b: in6_addr): in6_addr
extern function ipv6_mask_is_any(mask: in6_addr): bool
extern function scan_eth_addr(s: string): Option<eth_addr>
extern function json_string_escape(s: string): string

extern function ip_parse_masked(s: string): Either<string/*err*/, (ovs_be32/*host_ip*/, ovs_be32/*mask*/)>
extern function eth_addr_from_string(s: string): Option<eth_addr>
