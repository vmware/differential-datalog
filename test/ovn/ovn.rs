use std::iter::FromIterator;

fn types_map_extract_val_uuids<K: Val>(ids: &std_Map<K, types_uuid_or_string_t>) -> std_Map<K, types_uuid> {
    std_Map::from_iter(ids.x.iter().map(|(k,v)| (k.clone(), types_extract_uuid(v))))
}
fn types_set_extract_uuids(ids: &std_Set<types_uuid_or_string_t>) -> std_Set<types_uuid> {
    std_Set::from_iter(ids.x.iter().map(|x| types_extract_uuid(x)))
}
