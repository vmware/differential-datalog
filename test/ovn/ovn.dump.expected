ovn_nb_NB_Global:
ovn_nb.NB_Global{0,1,2,3,types.Some{59639315971019240521003784126651274906}}
ovn_nb.NB_Global{0,1,2,3,types.None{}}

